doc/man/man4/rmux.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the section entitled "GNU Free
.\" Documentation License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaning to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF
.\" SUCH CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log$
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore T
.R2
.so openss7.macros
.\"
.\"
.TH RMUX 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B rmux
\- Remote Multiplexing Driver for STREAMS
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.br
.B #include <ioctl.h>
.HP
.BI "int open(" rmux_name ", " o_flags );
.\"
.\"
.SH DESCRIPTION
.PP
.B RMUX
is a remote multiplexing driver using SCTP,
.[
rfc2960
.]
.[
rfc3309
.]
.[
rfc4460
.]
.BR sctp (4).
Its purpose is to make remote
.IR STREAMS (9)
devices appear to be local STREAMS devices, permitting distribution of protocol stacks over multiple
hosts in a fashion transparent to applications programs.  The
.B RMUX
driver uses SCTP,
.BR sctp (4),
as its transport protocol and a slim protocol to achieve transparency to the STREAMS user interface.
.PP
.B RMUX
uses a single SCTP association to the remote host to provide access to all of the STREAMS devices on
that host.
.B RMUX
is a multiplexing driver that links SCTP associations beneath it using a control Stream during the
configuration process.  Device numbers on the upper mutiplex (both major and minor) are configured
during the configuration process.
.PP
.B RMUX
is cleaner than SIGTRAN for use within distributed OpenSS7 protocol stack systems.  Instead
of being specific to a protocol layer service primitive interface,
.B RMUX
is general purpose and
can be use to export the User-Provider interface for _any_ protocol module.
.PP
Non-STREAMS based systems that have SCTP can open and manage STREAMS devices as long as they use
the
.B RMUX
protocol directly.
.PP
By chaining devices between hosts, it is possible to provide redundant paths to the same device.
This accomplished by opening the device multiple times via multiple systems.
.B RMUX
can be configured to understand that multiple upper multiplex Streams are related in this fashion
and can provide automatic failover from one path to another.  Because the acknowledged SCTP NPI
service is used, each encapsulated message can be marked with an identifier and the ultimate
receiving device can remove duplications.
.\"
.SS Initialization
.PP
At the local
.B RMUX
driver, the RMUX daemon listens for SCTP connections on a well-known port number.  For each
client connection configured the RMUX daemon forms an outgoing SCTP connection to the remote host.
For each client or server connection, the RMUX daemon sends an R_INIT_REQ message and awaits an
R_INIT_ACK message.  The R_INIT_REQ message provides information about the RMUX driver as follows:
.TP \w'\fIendianess\fR\(em'u
.I version
The version of the RMUX protocol.
.TP
.I endianess
The endianess of the host sending the R_INIT_REQ.
.PP
Sending and receiving
.I endianess
is important.  If there is an endian mismatch between the client and the server, full transparency
cannot be acheived.  Each driver opened must identify the format of the service primitive interface
for the driver to be able to perform the necessary conversion.  Conversion is always performed at
the server side of a connection (where the service interface can be defined).  When a Stream is
opened on the server, it pushes an endian conversion module before linking it under the server-side
.B RMUX
driver.
.\"
.SS Open and Close
.PP
When a user opens a configured minor device number, the local
.B RMUX
driver blocks in the
.BR qi_qopen (9)
procedures and communicates with the remote host on a linked SCTP Stream.
The remote
.B RMUX driver has a
daemon process on a control Stream that, upon an indication, opens the requested device,
links it under the
.B RMUX
driver at the remote end and communicates a return status to the requesting end.  The requesting end
wakes the sleeping
.BR qi_qopen (9)
process and returns the appropriate return value.
.PP
Subsequent opens of the local
.B RMUX
Stream that invoke the
.BR qi_qopen (9)
procedure of the local upper multiplex Stream are also communicated to the remote end and the remote
.B RMUX
driver mimics the Stream head open down the module stack and provides the appropriate return value.
.PP
When the user closes the driver,
.B RMUX
blocks the user in the
.BR qi_qclose (9)
procedure,
communicates the close to the remote end.  The daemon at the remote end unlinks and closes the
driver and communicates the close to the remote end.
.PP
Users cannot rely on the device numbering of opened Streams.
.PP
Messages passed to the Stream head that are absorbed by the Stream head (e.g.,
.BR M_ERROR (9),
.BR M_SIG "(9), " M_SETOPTS (9))
on open may be lost for the remote system.  Using knowledge about the Linux Fast-STREAMS
implementation it is possible to determine whether
.BR M_HANGUP "(9) or " M_SETOPTS (9)
as performed while linking.
.TP
.B First Open
.TP
\fIActions at the client \fBRMUX\fP driver:
.PP
The following actions are taken resulting from the first open of a device at the client
.B RMUX
driver:
.IP \(em \w'\(em\(em'u
When a local user process opens a configured minor device on the
.B RMUX
driver, resulting in the first open of the minor device, the
.B RMUX
driver
.BR qopen (9)
routine sends an
.B R_OPEN_REQ
message to the corresponding serving
.B RMUX
driver on the corresponding SCTP connection, and blocks the calling user awaiting a timeout, signal
or reception of optional
.B R_NORMAL_REQ
or
.B R_HIPRI_REQ
messages, followed by a
.B R_OPEN_CON
message confirming the open request.
.IP \(em
When an
.B R_OPEN_CON
message is received, or a timeout occurs, or a signal is received, the user process is awoken in the
.BR qopen (9)
call, and returns the appropriate error number.
.IP \(em
When awoken by a
.B R_OPEN_CON
message confirming successful open of the device at the serving
.B RMUX
driver, the message also includes the value of the effective
.I sflag
argument to
.BR qopen (9).
.IP ""
If the effective
.I sflag
argument is
.BR CLONEOPEN ,
a new minor unique device number is assigned before success is returned.
(Note that if no minor device numbers remain available for assignment, the client
.B RMUX
driver will have to issue a separate
.B R_CLOSE_REQ
primitive to close the Stream alreay open at the serving
.BR RMUX
driver.)
.IP ""
If the effective
.I sflag
argument is
.BR DRVOPEN ,
the current device number is used.

NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-04-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel20081029_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.

   - Modifications to build under `Fedora' `2.6.22.5-49' kernel.  These
     changes also support `2.6.22.9-91.fc7' kernel.  Modifications to
     build under `Fedora' `2.6.25-45.fc9' and `2.6.26.5-45.fc9' kernels.
     Documented `lib32gcc1' problem on Ubuntu.  Noted problem running
     under XEN kernels.  XEN kernels are not yet supported.  Added
     `MODULE_VERSION' to all modules and drivers.

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.  Improvements to common build
     process for embedded and cross-compile targets.  Cross-compile
     fixes (strap out `AC_FUNC_REALLOC' macro when cross-compiling).
     Conversion of RPM spec files to common approach for major
     subpackages.

     Build system now builds `yum(8)' repositories for RPMs and
     `apt-get(8)' repositories for DEBs.  Installation documentation
     has been updated to include details of repository install
     sourcesref.


src/drivers/rmux.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/*
 *  RMUX is a remote multiplexing driver using SCTP.  Its purpose is to make remote STREAMS devices
 *  appear to be local STREAMS devices, permitting distribution of protocol stacks over multiple
 *  hosts in a fashion transparent to applications programs.  The RMUX driver uses SCTP as its
 *  transport protocol and a slim protocol to achieve transparency to the STREAMS user interface.
 *
 *  RMUX uses a single SCTP association to the remote host to provide access to all of the STREAMS
 *  devices on that host.  RMUX is a multiplexing driver that links SCTP associations beneath it
 *  using a control Stream during the configuration process.  Device numbers on the upper mutiplex
 *  (both major and minor) are configured during the configuration process.
 *
 *  Open and Close
 *
 *  When a user opens one of these defined upper device numbers, RMUX communicates with the remote
 *  host whose daemon process opens the requested device, links it under the RMUX driver at the
 *  remote end and provides a return status to the user.  When the user closes the driver, RMUX
 *  communicates the close to the remote end.
 *
 *  I_LINK and I_PLINK
 *
 *  Local users can also link both remote and local Streams under a remote multiplexing driver.
 *  When linking a Stream, it is communicated to the remote RMUX during the linking process and the
 *  remote RMUX daemon opens a special minor device on its upper multiplex and links it under the
 *  remote driver.  When linking a remote Stream on the same host, the corresponding Stream on the
 *  remote RMUX lower multiplex is unlinked and linked under the appropriate driver.  When linking a
 *  remote Stream on a different host is the same a linking a local Stream.
 *
 *  I_PUSH and I_POP
 *
 *  Local users can push modules onto remote Streams; however, the module pushed is a local module.
 *
 *  STREAMS message handling.
 *
 *  All STREAMS messages are encapsulated and sent to the remote RMUX.  The multiple stream
 *  capability of SCTP is utilized fully.  When a remote Stream is established, an SID is assigned
 *  to each of queue band number for the Stream.  Normal priority messages (e.g, M_PROTO) are sent
 *  using ordered delivery on the SID that corresponds to the queue band.  High priority messages
 *  (e.g, M_PCPROTO) are sent using unordered (???) delivery on the SID that corresponds to high
 *  priority messages (likely SID 0).
 *
 *  Flow Control
 *
 *  Flow control along the Stream is accomplished using specialized RMUX messages that indicate the
 *  band for which flow control is effected.  The receiving RMUX driver places a specalized blocking
 *  message on the appropriate queue pair to establish backpressure.  Additional RMUX message are
 *  used for backenabling and also indicate the band for which the backenable has occurred
 *  (resulting in the removal of the blocking message from the associated queue band).  In this way,
 *  STREAMS flow control operates transparently.
 *
 *  A better approach to flow control would be to leave messages on the local sending queue until
 *  they are acknowledged received by the remote end (using the acknowledged data service of the NPI
 *  interface to SCTP).  This is cleaner in that it does not require blocking messages.  It does
 *  still require a back-enable message, but without band number.
 *
 *  Issues:
 *
 *  1. Module pushing is problematic in that there may be module mismatches between the local and
 *     remote system.  Where closer control is required a workaround is to use the autopush
 *     mechanism on the remote system instead (have modules autopushed on opened drivers instead
 *     of explicitly pushed by the local user).  Another workaround is to use a specialized set of
 *     R_PUSH, R_POP, input-output controls that mirror the I_PUSH, I_POP, input-output controls for
 *     STREAMS.  The problem with this later approach is that the user must be aware that an RMUX
 *     Stream is being used.  (Unless we provide a Stream head option to pass I_PUSH and I_POP
 *     input-output controls down to the driver.  This later approach is probably best as it also
 *     allows remote Streams on to be linked under multiplexing drivers on the same host.)
 *
 *     The remote system will have to unlink the Stream, push or pop the module and then
 *     relink it.
 *
 *  2. Numbering, flags and structure layout.  It is likely that the same STREAMS implementation is
 *     necessary on both the sending and receiving system.  In particular, the remote driver must
 *     understand 32-bit and 64-bit flags for input-output controls.  Where incompatibilities exist,
 *     the module or driver must refrain from using those facilities.
 *
 *  3. All users must properly handle M_ERROR and M_HANGUP messages.  If communication is lost to
 *     the remote RMUX, all corresponding local Streams will be hungup with M_HANGUP.
 *
 *  4. Modules and drivers cannot have knowledge of, nor rely upon, duplicated message blocks,
 *     special alignment or "hidden" data in a message block between modules.  (Not much of an issue
 *     as they shouldn't anyway.)
 *
 *  5. Local Streams representing remote devices will always have service procedures.
 *
 *  6. Modules and drivers cannot have knowledge of, nor rely upon, the context in which messages
 *     are passed to neighbouring modules and drivers.
 *
 *  7. Users cannot rely on the device numbering of opened Streams.
 *
 *  8. Messages passed to the Stream head that are absorbed by the Stream head (e.g., M_ERROR,
 *     M_SIG, M_SETOPTS) on open may be lost for the remote system.  Using knowledge about the Linux
 *     Fast-STREAMS implementation it is possible to determine whether M_HANGUP or M_SETOPTS as
 *     performed while linking.
 *
 *  Advantages:
 *
 *  RMUX is cleaner than SIGTRAN for use within distributed OpenSS7 protocol stack systems.  Instead
 *  of being specific to a protocol layer service primitive interface, RMUX is general purpose and
 *  can be use to export the User-Provider interface for _any_ protocol module.
 *
 *  Non-STREAMS based systems that have SCTP can open and manage STREAMS devices as long as they use
 *  the RMUX protocol directly.
 *
 *  By chaning devices between hosts, it is possible to provide redundant paths to the same device.
 *  This accomplished by opening the device multiple times via multiple systems.  RMUX can be
 *  configured to understand that multiple upper multiplex Streams are related in this fashion and
 *  can provide automatic failover from one path to another.  Because the acknowledged SCTP NPI
 *  service is used, each encapsulated message can be marked with an identifier and the ultimate
 *  receiving device can remove duplications.
 *
 *  RMUX Protocol:
 *
 *  The RMUX protocol communicates by passing protocol messages to peer RMUX drivers on SCTP stream
 *  number 0.
 *
 *  R_OPEN    - open a device
 *  R_CLOSE   - close a device
 *  R_PUSH    - push a module
 *  R_POP     - pop a module
 *  R_LINK    - link a device
 *  R_PLINK   - permanent link a device
 *  R_UNLINK  - unlink a device
 *  R_PUNLINK - permanent unlink a device
 *  R_ENABLE  - enable a queue
 *  R_NORMAL  - pass a normal priority message
 *  R_HIPRI   - pass a high priority message
 *
 *  R_OPEN_REQ/IND: - open a device
 *
 *	32-bit Stream Identifier (local device number)
 *	32-bit open flags
 *	structured credentials
 *	32-bit length of device name
 *	32-bit offset of device name
 *	followed by device name
 *
 *  R_OPEN_RES/CON: - open a device
 *
 *	32-bit Stream Identifier (local device number)
 *	32-bit return value (negative error)
 *
 *  R_CLOSE_REQ/IND: - close a device
 *
 *	32-bit Stream Identifier (local device number)
 *	structured credentials
 *
 *  R_CLOSE_RES/CON: - close a device
 *
 *	32-bit Stream Identifier (local device number)
 *	32-bit return value (negative error)
 *
 *  R_PUSH_REQ/IND: - push a module
 *
 *	32-bit Stream Identifier (local device number)
 *	structured credentials
 *	32-bit length of module name
 *	32-bit offset of module name
 *	followed by module name
 *
 *  R_PUSH_RES/CON: - push a module
 *
 *	32-bit Stream Identifier (local device number)
 *	32-bit return value (negative error)
 *
 */

/*
 * The connection manager responds with an R_OPEN_RES primitive in response to an R_OPEN_IND
 * primitive.  The primitive is passed directly in an R_OPEN_CON message to the remote user.
 */
static int
r_open_res(struct rmux *rm, queue_t *q, mblk_t *mp)
{
}

static int
r_send_open_con(struct rmux *rm, queue_t *q, mblk_t *mp)
{
}

/*
 * When an R_OPEN_REQ message is received for a Stream has not already been linked by the device, it
 * issues an R_OPEN_IND service primitive to the control Stream and awaits a R_OPEN_RES primitive.
 * If no control Stream currently exists, an error code [ENXIO] is returned to the remote end in an
 * R_OPEN_RES message.
 *
 * The connection manager on the control Stream, upon receipt of an R_OPEN_IND primitive should open
 * the appropriate Stream, and when successful, link it under the multiplexing driver, and return an
 * R_OPEN_RES primitive with the multiplex identifier to associate it with the Stream.  When an
 * error occurs during the open of the Stream, an R_OPEN_RES primitive with the error is returned.
 */
static int
r_recv_open_req(struct rmux *rm, queue_t *q, mblk_t *mp)
{
}

/*
 * When an R_REOPEN_REQ message is received for a Stream that has already been linked, the driver
 * simlulates a Stream head module stack open and returns the return value in an R_REOPEN_CON
 * message.
 */
static int
r_recv_reopen_req(struct rmux *rm, queue_t *q, mblk_t *mp)
{
}

#if 0
static int
r_m_read(queue_t *q, mblk_t *mp)
{
	switch (DB_TYPE(mp)) {
	case M_READ:		/* bad, contains unsigned long *//* FIXME */
		break;
	case M_DELAY:		/* int value */
		if (mp->b_wptr >= mp->b_rptr + sizeof(uint32_t))
			*(uint32_t *) mp->b_rptr = cpu_to_be32(*(uint32_t *) mp->b_rptr);
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

/* 40-bytes for 32-bit, 58-bytes for 64-bits and a fastbuf is not less than 64 bytes */
struct stroptions64 {
	uint32_t so_flags;
	int16_t so_readopt;
	uint16_t so_wroff;
	int64_t so_minpsz;
	int64_t so_maxpsz;
	uint64_t so_hiwat;
	uint64_t so_lowat;
	uint8_t so_band;
	uint8_t so_filler[3];
	uint16_t so_erropt;
	int64_t so_maxblk;
	uint16_t so_copyopt;
	uint16_t so_wrpad;
};

static int
r_m_setopts(queue_t *q, mblk_t *mp)
{
	switch (DB_TYPE(mp)) {
	case M_SETOPTS:	/* really bad, contains ssize_t and size_t *//* FIXME */
		break;
	case M_PCSETOPTS:	/* really bad, contains ssize_t and size_t *//* FIXME */
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

static int
r_m_ioctl(queue_t *q, mblk_t *mp)
{
	switch (DB_TYPE(mp)) {
	case M_IOCTL:		/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_IOCACK:		/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_IOCNAK:		/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_COPYIN:		/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_COPYOUT:	/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_IOCDATA:	/* really bad, contains long and ulong and pointers *//* FIXME */
		break;
	case M_CTL:		/* bad, format unknown */
		break;
	case M_PCCTL:		/* bad, format unknown */
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

static int
r_m_letsplay(queue_t *q, mblk_t *mp)
{
	switch (DB_TYPE(mp)) {
	case M_BACKWASH:
		/* bad, contains attache M_DATA, M_PROTO or M_PCPROTO, but we should never get one
		   of this because we interrupt M_LETSPLAY */
		freemsg(mp->b_cont);
		mp->b_datap->db_type = M_DONTPLAY;
		break;
	case M_LETSPLAY:	/* refuse */
		mp->b_datap->db_type = M_DONTPLAY;
		qreply(q, mp);
		return (0);
	case M_DONTPLAY:
		break;
	case M_BACKDONE:	/* complicated but we shouldn't get this */
		mp->b_datap->db_type = M_DONTPLAY;
		break;
	default:
		break;
	}
	putnext(q, mp);
	return (0);
}

static int
r_m_other(queue_t *q, mblk_t *mp)
{
	switch (DB_TYPE(mp)) {
	case M_PASSFP:
		/* really bad, we can't do this */
		freemsg(mp);
		return (0);
	case M_READ:
	case M_DELAY:
		return cd_m_read(q, mp);
	case M_SETOPTS:
	case M_PCSETOPTS:
		return cd_m_setopts(q, mp);
	case M_IOCTL:
	case M_IOCACK:
	case M_IOCNAK:
	case M_COPYIN:
	case M_COPYOUT:
	case M_IOCDATA:
	case M_CTL:
	case M_PCCTL:
		return cd_m_ioctl(q, mp);
	case M_BACKWASH:
	case M_LETSPLAY:
	case M_DONTPLAY:
	case M_BACKDONE:
		return cd_m_letsplay(q, mp);
	default:
	case M_BREAK:
	case M_EVENT:		/* osbolete anyway */
	case M_SIG:		/* unsigned char - ok */
	case M_RSE:		/* bad */
	case M_TRAIL:		/* bad */
	case M_PCSIG:		/* just a char like M_SIG */
	case M_FLUSH:		/* just bytes */
	case M_STOP:		/* nothing */
	case M_START:		/* nothing */
	case M_HANGUP:		/* nothing */
	case M_ERROR:		/* just bytes */
	case M_PCRSE:		/* bad */
	case M_STOPI:		/* bad, format unknown */
	case M_STARTI:		/* bad, format unknown */
	case M_PCEVENT:	/* osbolete anyway */
	case M_UNHANGUP:	/* nothing */
	case M_NOTIFY:		/* nothing */
		break;
	}
	putnext(q, mp);
	return (0);
}
#endif

/*
 * There are two types of upper Streams: control Streams and user Streams.  The control Stream is a
 * clone minor device that can only be opened once.  User Streams are either clone or non-clone
 * Streams (depending on the response from the remote end).  User Streams must be configured by the
 * control Stream before they can be successfully opened.
 *
 * When configured, the minor device is assigned a lower mux SCTP transport Stream, device
 * identifier that identifies the device to the remote system, and a logical path on the remote
 * system.  The logical path might correspond to the actual filesystem path on the remote system to
 * use to open the device, or might be a logical reference.
 *
 * For a configured minor device, the r_qopen() procedure passes an R_OPEN_REQ message on the SCTP
 * stream to the remote system and blocks awaiting a response, timeout or loss of communications.
 * The R_OPEN_REQ message contains the 32-bit local device instance identifier (derived from the q
 * pointer), the logical path name (representing the device opened), the oflags, and the
 * credientials of the opening user.
 *
 * If STREAMS messages were passed to the Stream head as part of the remote open procedure, a
 * R_NORMAL_IND or R_HIPRI_IND message might arrive before the R_OPEN_CON, in which case the
 * deencapsulation of those messages are passed to the local Stream head (likely M_SETOPTS).
 *
 * If a timeout or loss of communications on the SCTP Stream occurs before the open completes, an
 * appropriate error code is returned and the open call fails.
 *
 * A response R_OPEN_CON message will contain the device instance identifier (as was provided in the
 * R_OPEN_REQ message), applied STREAMS flags (sflag) indicating DRVOPEN or CLONEOPEN, and a return
 * value indicating either success or an error.  When an error occurs, the associated error return
 * code is returned and no other action is taken.  When successful, the returned sflag is examined
 * to determine whether the remote end assigned a unique minor or not.  When sflag is CLONEOPEN, the
 * open will complete as though CLONEOPEN was specified to this call; when DRVOPEN, as though
 * DRVOPEN was specified on the r_qopen() call.
 *
 * When a user Stream is already opened for the first time and a subsequent open call occurs,
 * Another R_OPEN_REQ message is passed to the remote end and the calling process is blocked
 * awaiting the R_OPEN_CON.  The return value from the R_OPEN_CON is again returned.
 */
static streamscall int
r_qopen(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
}

/*
 * There are two types of upper streams: control Streams and user Streams.  Control Streams are
 * closed as normal.  User Streams communicate the device arguments (device number and credentials)
 * in a R_CLOSE_REQ message to the remote end and blocks the user awaiting a response (or timeout or
 * communications failure).
 */
static streamscall int
r_qclose(queue_t *q, dev_t *devp, cred_t *crp)
{
}
